Add unittest for event_watch_fn callback

andreas-bok-sociomantic opened the pull request:
Tests that event_watch_fn is set correctly
and called with valid arguments

andreas-bok-sociomantic committed:
Add unittest for event_watch_fn callback

Tests that event_watch_fn is set correctly
and called with valid arguments

andreas-bok-sociomantic commented:
Added to `v5.2.x` release since it doesn't add to the existing api

ben-palmer-sociomantic commented on tests/unit/event_callback.c:
```diff
@@ -0,0 +1,95 @@
+ *
+ * Drizzle Client & Protocol Library
+ *
+ * Copyright (C) 2016 Andreas Bok Andersen (andreas.bok@sociomantic.com)
```
Is this the correct copyright notice to have here, or should it be our standard one?

leandro-lucarella-sociomantic replied:
Authorship and copyright are not the same, the author should be @andreas-bok-sociomantic but the copyright holder should be Sociomantic Labs GmbH, as long as this work has been done inside Sociomantic :)

ben-palmer-sociomantic commented on tests/unit/event_callback.c:
```diff
@@ -0,0 +1,95 @@
+  ASSERT_TRUE((events != -1));
+  int* cxt_ptr = (int*) drizzle_context(con);
+  ASSERT_NOT_NULL_(cxt_ptr, "Drizzle connection context is null");
+  ASSERT_EQ_( 1, cxt_ptr[0], "Invalid drizzle context: Got '%d', Expected '1'",
```
Extra space between `(` and `1`.

ben-palmer-sociomantic commented:
I had a query on the copyright notice but otherwise LGTM.

andreas-bok-sociomantic commented:
- updated based on comments from @ben-palmer-sociomantic 
- fixed formatting of 80+ lines

```diff 
diff --git a/tests/unit/event_callback.c b/tests/unit/event_callback.c
index af70626..6abfbb8 100644
--- a/tests/unit/event_callback.c
+++ b/tests/unit/event_callback.c
@@ -2,7 +2,7 @@
  *
  * Drizzle Client & Protocol Library
  *
- * Copyright (C) 2016 Andreas Bok Andersen (andreas.bok@sociomantic.com)
+ * Copyright (C) 2016 Sociomantic Labs GmbH
  * All rights reserved.
  *
  * Redistribution and use in source and binary forms, with or without
@@ -44,13 +44,15 @@
 #include <string.h>
 #include <stdlib.h>
 
-extern drizzle_return_t drizzle_event_callback(drizzle_st *con, short events, void *context);
-extern drizzle_return_t drizzle_event_callback(drizzle_st *con, short events, void *context)
+extern drizzle_return_t drizzle_event_callback(drizzle_st *con, short events,
+  void *context);
+extern drizzle_return_t drizzle_event_callback(drizzle_st *con, short events,
+  void *context)
 {
   ASSERT_TRUE((events != -1));
   int* cxt_ptr = (int*) drizzle_context(con);
   ASSERT_NOT_NULL_(cxt_ptr, "Drizzle connection context is null");
-  ASSERT_EQ_( 1, cxt_ptr[0], "Invalid drizzle context: Got '%d', Expected '1'",
+  ASSERT_EQ_(1, cxt_ptr[0], "Invalid drizzle context: Got '%d', Expected '1'",
     cxt_ptr[0]);
 
   cxt_ptr = (int*)context;
@@ -82,7 +84,8 @@ int main(int argc, char *argv[])
   drizzle_set_event_watch_fn(con, drizzle_event_callback, (void*)&cxt_b);
 
   drizzle_return_t driz_ret= drizzle_connect(con);
-  SKIP_IF_(driz_ret == DRIZZLE_RETURN_COULD_NOT_CONNECT, "%s", drizzle_strerror(driz_ret));
+  SKIP_IF_(driz_ret == DRIZZLE_RETURN_COULD_NOT_CONNECT, "%s",
+    drizzle_strerror(driz_ret));
   ASSERT_EQ_(DRIZZLE_RETURN_OK, driz_ret, "%s(%s)", drizzle_error(con),
     drizzle_strerror(driz_ret));

```

andreas-bok-sociomantic commented:
fixed test fail and added description of assert 

```
  // The drizzle_event_callback is invoked by drizzle_set_events and
  // drizzle_set_revents with io changes.
  // Internally the changes occur during the following steps:
  //
  // 1. init drizzle connection (drizzle_connect)
  // 2. while connecting to server (drizzle_state_connecting)
  // 3. performing server handshake (drizzle_state_handshake_client_write)
  // 4. closing connection (drizzle_quit)
  //
  // Please note that stepping through the test with 'make gdb-event_callback'
  // can cause the test to fail
  ASSERT_EQ_(4, cxt_b, "Invalid drizzle context: Got '%d', Expected '2'", cxt_b);
```

ben-palmer-sociomantic commented on tests/unit/event_callback.c:
```diff
@@ -0,0 +1,109 @@
+  // 1. init drizzle connection (drizzle_connect)
+  // 2. while connecting to server (drizzle_state_connecting)
+  // 3. performing server handshake (drizzle_state_handshake_client_write)
+  // 4. closing connection (drizzle_quit)
```
Should the above documentation should be here or at a different part of the test. I also find the wording of step 2 and 3 a bit confusing so perhaps you could try and re-write them.

andreas-bok-sociomantic replied:
> Should the above documentation should be here or at a different part of the test. 
> The other unittests don't have a description of the test apart from the assert statement messages. 
> I guess it could be moved to the file header

andreas-bok-sociomantic commented:
updated based on comments from @ben-palmer-sociomantic 
- rephrased the unittest description 
- fixed wrong assert error message 

Not sure about the location of the description though. 

```diff
diff --git a/tests/unit/event_callback.c b/tests/unit/event_callback.c
index 79f5155..e512c83 100644
--- a/tests/unit/event_callback.c
+++ b/tests/unit/event_callback.c
@@ -92,18 +92,19 @@ int main(int argc, char *argv[])
   driz_ret= drizzle_quit(con);
   ASSERT_EQ_(DRIZZLE_RETURN_OK, driz_ret, "%s", drizzle_strerror(driz_ret));
 
-  // The drizzle_event_callback is invoked by drizzle_set_events and
-  // drizzle_set_revents with io changes.
-  // Internally the changes occur during the following steps:
+  // Check that the drizzle_event_callback was called. It is done indirectly by
+  // asserting that cxt_b has been incremented 4 times.
+  // The increments of cxt_b correspond to the following state changes:
   //
-  // 1. init drizzle connection (drizzle_connect)
-  // 2. while connecting to server (drizzle_state_connecting)
-  // 3. performing server handshake (drizzle_state_handshake_client_write)
-  // 4. closing connection (drizzle_quit)
+  // 1. when drizzle connection is initied by calling drizzle_connect
+  // 2. while connecting to server
+  // 3. when server handshake succeeds
+  // 4. when the drizzle_st connection is closed by calling drizzle_quit
   //
   // Please note that stepping through the test with 'make gdb-event_callback'
   // can cause the test to fail
-  ASSERT_EQ_(4, cxt_b, "Invalid drizzle context: Got '%d', Expected '2'", cxt_b);
+  ASSERT_EQ_(4, cxt_b, "Unexpected number of event callbacks: Got '%d', \
+    Expected '4'", cxt_b);
 
   return EXIT_SUCCESS;
 }
```

ben-palmer-sociomantic commented:
LGTM, the documentation location is fine I just misunderstood what was happening.

ben-palmer-sociomantic commented:
This pull request has been rebased via `git hub pull rebase`. Original pull request HEAD was 1b4d221988f6c5ea4d9af0147fc8c014523bffce, new (rebased) HEAD is ffded993548a7cc2e4eeeb5da474a40c3dd917f1
